db/db.go

```diff
@@ -0,0 +1,75 @@
+package db
+
+/*
+
+client<->entity relationship:
+
+get/set/delete accept the Path() argument from entity to perform key/value
+lookup. Get and set respectively manage Read() and Write() from entity.
+*/
+
+// Client provides a method of hooking into our data stores. It processes
+// entities (see below) and their pipeline between them and k/v stores such as
+// etcd and consul.
+type Client interface {
+	// Get the entity. Will copy the contents of the JSON into the entity for
+	// use.
+	Get(Entity) error
+	// Set the entity. Will persist the entity to etcd based on the path it
+	// provides. (See Path())
+	Set(Entity) error
```

Comment:
That’s a good question. Currently the usage patterns are enforced in 
code, which is necessary only for volumes at the moment.

The problem with “add vs set” however is really about what is adding 
and what is setting. In a distributed environment there could be two 
adds that happen at once or two sets. The semantics here get kind of 
wacky, but it could be done.

All that said until the usage patterns need it I would like to avoid 
amending this behavior.

On 15 Aug 2016, at 22:07, Vikrant Balyan wrote:

>> +client<->entity relationship:
>> +
>> +get/set/delete accept the Path() argument from entity to perform 
>> key/value
>> +lookup. Get and set respectively manage Read() and Write() from 
>> entity.
>> +*/
>> +
>> +// Client provides a method of hooking into our data stores. It 
>> processes
>> +// entities (see below) and their pipeline between them and k/v 
>> stores such as
>> +// etcd and consul.
>> +type Client interface {
>> +	// Get the entity. Will copy the contents of the JSON into the 
>> entity for
>> +	// use.
>> +	Get(Entity) error
>> +	// Set the entity. Will persist the entity to etcd based on the 
>> path it
>> +	// provides. (See Path())
>> +	Set(Entity) error
>
> @erikh question: are there any stores which differentiate Add and Set 
> ? Add - first time, Set - change(could be a subset of the entry)? If 
> so, we need to design for them here.
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/contiv/volplugin/pull/408/files/1a3157f1b1181a25f0c3c2e5792ffecb6ca6e07d#r74876708